uktrade/defend-data-capture
# Configure app for deployment to cloud foundry

This PR makes the app ready to be deployed on cloud foundry.

The two main files needed for deployments are `manifest.yml` and `Procfile`.

- `manifest.yml` specifies the buildpacks which should be used for the application. The node buildpack comes first, as the main buildpack (which corresponds to the start command in the Procfile) should come second.
- `Procfile` specifies the start command which should be run on the python app. This applies migrations to the db and maps the gunicorn server to the expected port (it automatically runs on 8080 in cloud foundry)

The PR also does the following:

- Sets the python and node versions to match those compatible with their [buildpacks](https://docs.cloudfoundry.org/buildpacks/)
- Moves the requirements.txt file to the root of the project
- Configure settings for static files, so that when `collectstatic` is run by the python buildpack, the static files are accessible at `/static/` in the root of the project.
- Adds the Whitenoise package in order to serve static files on the same WSGI (gunicorn) server as the application ([docs here](https://readme.trade.gov.uk/docs/patterns/django-static-files.html?highlight=whitenoise#architecture-for-service-of-static-files))
- Installs gunicorn for creation of servers in live environments
- Sets `ALLOWED_HOSTS` to come from environment variables (with default as localhost) - this will allow us to avoid source controlling the names of our URLs.

**To test**
You can manually push the app by doing the following:

1. Check out this branch
2. Assuming you have the cloud foundry CLI installed, log in via sso - `cf login --sso`
3. Choose the `dit-staging` org and the `supply-chain-resilience-dev` space
4. Ensuring you're in the root of the project in your terminal, run `cf push update-supply-chain-information-dev`
5. The deployment takes around 6 minutes

**TODO**
- A webops engineer will create a jenkins job to set up automatic deployments to our dev/staging environments whenever code is merged into main, and manual ones for production.
